Add Show Sample Shape to ADS right-click menu

DanielMurphy22 opened the pull request:
**Description of work.**
 
Added right-click option to the ADS to allow the Sample Shape to be plotted.
This will plot all sample, container and component shapes attached to a workspace,
add a beam direction arrow in black and  if there is an attached crystal lattice (UB Matrix) then the real+reciprocal lattice vector arrows in rgb. These plots should also rename with the workspace, replot when the workspace is updated, in case SetSample or SetGoniometer was called, and close when the workspace is deleted.

Note two software behaviours that are not genuine problems but I came across in my own testing: Containers are not rotated by the goniometer. You may also notice that if a workspace has a container and you add just a sample then it will keep the container!

**Report to:** Pas and Spencer

**To test:**

- Review documentation MeshPlotHelp.rst which has been updated. It can be accessed via the `?` toolbar button on a SampleShapeplot (if you have built `docs-qthelp` and `docs-qtassistant`)
- Create a workspace with a sample and UB matrix:
```
cuboid = " \
<cuboid id='some-cuboid'> \
<height val='2.0'  /> \
<width val='2.0' />  \
<depth  val='0.2' />  \
<centre x='10.0' y='10.0' z='10.0'  />  \
</cuboid>  \
<algebra val='some-cuboid' /> \
"

ws_shape = CreateSampleWorkspace()
SetGoniometer(ws, Axis0="120,0,0,1,1")
SetSample(ws, Geometry={'Shape': 'CSG', 'Value': cuboid})
SetUB(ws, a=1, b=1, c=2, alpha=90, beta=90, gamma=60)
```
- Check you get the black beam arrows and the coloured lattice vector arrows.
- Run SetSample for one of the examples below (on the same workspace) and the plot should update in the same figurewindow as before.
- Right-click on the workspace and rename it, the plot should update with the new name.
- Delete the workspace and the plot should close.
- Create another workspace and Show the Sample Shape multiple times. Check they all the plots get renamed/close when the workspace is renamed/deleted.


- Use combinations of the below SetSample calls for general messing (more example code here: https://docs.mantidproject.org/nightly/algorithms/SetSample-v1.html#algm-setsample):


Just sample:
```
CreateWorkspace(OutputWorkspace="ws_shape", DataX=[1, 1], DataY=[2, 2])
merge_xml = ' \
<cylinder id="stick"> \
<centre-of-bottom-base x="-0.5" y="0.0" z="0.0" /> \
<axis x="1.0" y="0.0" z="0.0" />  \
<radius val="0.05" /> \
<height val="1.0" /> \
</cylinder> \
\
<sphere id="some-sphere"> \
<centre x="0.7"  y="0.0" z="0.0" /> \
<radius val="0.2" /> \
</sphere> \
\
<rotate-all x="90" y="-45" z="0" /> \
<algebra val="some-sphere (: stick)" /> \
'
SetSample("ws_shape", Geometry={'Shape': 'CSG', 'Value': merge_xml})

```
Sample and Container:
```
CreateWorkspace(OutputWorkspace=workspace_name, DataX=[1, 1], DataY=[2, 2])
SetSample(workspace_name,
          Geometry={'Shape': 'Cylinder', 'Height': 4.0,
                    'Radius': 2.0, 'Center': [0., 0., 0.]},
          Material={'ChemicalFormula': '(Li7)2-C-H4-N-Cl6',
                    'NumberDensity': 0.1},
          ContainerGeometry={'Shape': 'HollowCylinder', 'Height': 4.0,
                             'InnerRadius': 2.0, 'OuterRadius': 2.3,
                             'Center': [0., 0., 0.]},
          ContainerMaterial={'ChemicalFormula': 'Al',
                             'NumberDensity': 0.01})
```
Just Container:
```
CreateWorkspace(OutputWorkspace=workspace_name, DataX=[1, 1], DataY=[2, 2])

SetSample(workspace_name,
          ContainerGeometry={'Shape': 'HollowCylinder', 'Height': 4.0,
                             'InnerRadius': 2.0, 'OuterRadius': 2.3,
                             'Center': [0., 0., 0.]},
          ContainerMaterial={'ChemicalFormula': 'Al',
                             'NumberDensity': 0.01})
```
Sample, Container and Components:
```
workspace = CreateWorkspace(OutputWorkspace="ws_shape", DataX=[1, 1], DataY=[2, 2])
LoadInstrument(Workspace=workspace, RewriteSpectraMap=True, InstrumentName="Pearl")
SetSample(workspace, Environment={'Name': 'Pearl'})
```

To test an instrument reference frame with pointing up Z and along beam X (e.g. POLREF) use this hack:
```
workspace = CreateWorkspace(OutputWorkspace="ws_shape", DataX=[1, 1], DataY=[2, 2])
LoadInstrument(Workspace=workspace, RewriteSpectraMap=True, InstrumentName="Polref")
SetSample(workspace, Environment={'Name': 'Pearl'})
```


Fixes #34563 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DanielMurphy22 committed:
Check workspace for valid sample shape

DanielMurphy22 committed:
Check shape valid by mesh and plot

DanielMurphy22 committed:
Plot Container and Components

DanielMurphy22 committed:
Set axes limits equal and by largest mesh limits

DanielMurphy22 committed:
Separate plot container from plot components

DanielMurphy22 committed:
Move plotting decorating code to separate funcitons

DanielMurphy22 committed:
validate container and components

DanielMurphy22 committed:
Add arrows for beam and lattice vectors

DanielMurphy22 committed:
Hook up Plotting.SampleShape to ADS right-click menu

Add to workspacewidget tests

DanielMurphy22 committed:
Support limits for all 3 dimensions

DanielMurphy22 committed:
Update colors and opacity(alpha)

rgb for sample, container, components

DanielMurphy22 committed:
Update warning message

DanielMurphy22 committed:
Add ADSobserver and FigureManagerObserver to SampleshapePlot

DanielMurphy22 committed:
Set window title and correct observers

DanielMurphy22 committed:
Prevent a plot being created for an  invalid sample shape

eg EMU data in TrainingCourseData

DanielMurphy22 committed:
Update cppcheck suppressions file

DanielMurphy22 committed:
set projection type to ortho

DanielMurphy22 committed:
restructure create_plot() and add ads observer tests

DanielMurphy22 committed:
Add release notes

DanielMurphy22 committed:
Update Sample Shape plot documentation

DannyHindson commented on docs/source/plotting/MeshPlotHelp.rst:
```diff
@@ -20,136 +18,38 @@
-   axes.add_collection3d(mesh_polygon)
+To quickly plot all Sample, Container and Component shapes attached to a workspace, right-click on it
+in the Workspaces Toolbox and select ``Show Sample Shape``. A black arrow will be added for the beam direction
+and, if there is a UB matrix, colored arrows for the real (solid line) and reciprocal (dashed line) lattice vectors.
```
Just wondering if it's meaningful to define which colours are used for which lattice vectors? So the first reciprocal lattice vector is the one associated with the h. So something like red=h, green=k, blue=l. But not sure if there's other better terminology here. Perhaps the first reciprocal lattice vector is commonly called a1 or something similar. Perhaps check with Richard

DanielMurphy22 replied:
I think you're right I should call the reciprocal b1, b2, b3 and real a1, a2, a3.

DannyHindson commented on docs/source/plotting/MeshPlotHelp.rst:
```diff
@@ -20,136 +18,38 @@
-   plt.show()
+Scripting
+#########
+Here are some examples of what you can achieve by plotting the sample shapes with a script. Click on ``Source Code``
```
Where is the source code link you can click on? Wasn't seeing it in my help page

DanielMurphy22 replied:
Inside the qthelp docs there is no source code link but if you view the page in a browser then the source code link is just above the plot. What do you think I should do?
![Screenshot from 2022-11-07 14-43-09](https://user-images.githubusercontent.com/55980573/200338288-efb7b0d0-ed3d-45c7-a6a7-9b546532c0b4.png)

DannyHindson replied:
OK I'd not noticed that difference between the qt and html docs before. I think it's probably OK to just clarify that the source will only be present if the user is viewing the html version of the docs. 

If you don't like that option then you could see if it's possibly to include your own hand made hyperlink to a separate file containing the python code source code. Not sure if that would involve duplicating the python or whether the plot could be generated from the separate file.

DannyHindson commented on qt/python/mantidqt/mantidqt/plotting/sample_shape.py:
```diff
@@ -0,0 +1,424 @@
+
+def calculate_lattice_vectors(workspace):
+    ub_matrix = np.array(workspace.sample().getOrientedLattice().getUB())
+    hkl = np.array([[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]])
```
Perhaps worth a comment here that the lattice vectors are in columns. Not sure the numpy syntax helps where because it suggests they're in rows. Later on you pull out individual columns when plotting the arrows

DanielMurphy22 replied:
This code is from the original scripts, so I was looking back at my Slack convo with Richard. And he had mentioned that the lattice vectors such as b1 are column vectors. But I guess I was meaning hkl to be rows. Is there a better way I can represent the code?

DannyHindson commented on qt/python/mantidqt/mantidqt/plotting/sample_shape_ads_observer.py:
```diff
@@ -0,0 +1,84 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 202 ISIS Rutherford Appleton Laboratory UKRI,
```
I'm getting pedantic. Small typo in the year here

DannyHindson reviewed (changes requested):
This is great! I tried plotting the PEARL sample + environment by loading a PEARL file (eg PEARL00073987.raw), running Sample shape with the environment parameter set to {"Name": "PEARL"} and it just works

As discussed separately - if you could determine the up direction in the plots from the frame object attached to the workspace that would be good.

A few other things:
1) I think that if the lattice vectors are outside of the axes ranges then the arrows are not getting shown. So in the flat plate (cuboid) example there is a blue reciprocal lattice vector with values ~ [-0.4,-0.25,0] but it's not getting shown at all. I got the vector details from debugging the sample_shape.py code
2) There was one small problem with my PEARL example but it relates to a pre-existing bug\problem rather than your changes on this PR. You get a warning\error reported on running SetSample from the algorithm dialog with the environment parameter I mention above ("The can has a fixed sample shape that cannot be adjusted using the Geometry parameter"). This can be fixed by changing the function SetSample::setSampleShape. The `if(args)` statements aren't quite correct. I'm wondering if we can include this in this PR - let's discuss
3) if a workspace doesn't have a sample shape I see that you write a warning. If you've got lots in your log already or have other windows open this isn't always totally obvious. Is it possible to hide the "show sample shape" right click option if there's no shape? I'm guessing perhaps not very easy because you'd need to react to someone adding a sample shape to a ws already in the ADS but thought I'd check. Fine to leave if this is tricky.

DanielMurphy22 commented on docs/source/plotting/MeshPlotHelp.rst:
```diff
@@ -20,136 +18,38 @@
-**Mesh Plots can only be accessed with a script, not through the Workbench interface**
+|
+|
+
```
Sorry yes this looks a little silly, so I'll remove the spacers I added!

DanielMurphy22 commented:
Other thing replies:
1. I think the blue reciprocal arrow lies exactly on top of the blue real arrow (their vectors are the same ratio). If you change the alpha value in SetUB from 90 then these 2 blue arrows start to separate: (e.g.`SetUB(ws, a=1, b=1, c=3, alpha=85, beta=90, gamma=60)`)
2. We should discuss!
3. That's a good idea, but sadly it wouldn't work very nicely. I could put an if statement into the workspacewidget to check there is a sample or even a shape. But I think that would be useless as even CreateWorkspace() has a sample shape. At the time I get the Mesh, I check it isn't meaningless (like an empty array). But I don't want to generate a mesh when the right-click ADS menu is generated.

DanielMurphy22 committed:
Address requested changes

DanielMurphy22 commented:
> I had another look at the "y up" behaviour following your latest changes. How about doing something like this:
> 
> ```
> plot_axes.view_init(vertical_axis = 'y')
> ```
> 


This is really good! vertical_axis is much better than roll as it sets the axes ticks nicely! vertical_axis works well in MPL 3.5. Sadly `axes.set_aspect('equal')` is only available in MPL 3.6 and so using `current_axes.set_box_aspect((1, 1, 1))` instead.

DanielMurphy22 committed:
Set aspect like equal and set vertical_axis

fix SetSample::setSampleShape with environment only from algm Dialog

DannyHindson reviewed (changes requested):
Just retested, looking good. A few minor things:
- [X] Can the source code examples be updated to plot the Y axis up as well? Also the proj_type setting to 'ortho' isn't in the source code examples - although lost track of whether that's the key thing on getting a square\equal plot
- [x] Could the reciprocal space arrows be made a slightly different length to the real space ones so the case where two of them are in the same direction displays a bit nicer? If you could see the two blue arrow heads that would be nice. Just tried doing the following and it seems to work OK if scale one type by 0.8:
```
def add_arrow(ax, vector, origin=None, factor=None, color='black', linestyle='-', scale_factor=1.0):
...
vector_norm = vector * scale_factor / np.linalg.norm(vector)
```
- [ ] Running SetSample twice on the same workspace doesn't update the shape on a plot eg initially flat plate and then change to sphere. Perhaps not essential right now but the specific use case where this might be useful is where a single crystal user changes the sample orientation in a script window and wants to see the sample plot update. I'm not referring to the changing the goniometer angles here - just the orientation without a goniometer.
- [x] I noticed that running SetGoniometer after SetSample doesn't update the sample orientation.   Perhaps it should. Not related to this PR though...

DanielMurphy22 commented:
> I noticed that running SetGoniometer after SetSample doesn't update the sample orientation. Perhaps it should. Not related to this PR though...
I agree that setgoniometer doesn't kick in until there is a new setsample call. I agree that we should think about that separately and whether the current behaviour is intended!

DanielMurphy22 commented:
> Running SetSample twice on the same workspace doesn't update the shape on a plot eg initially flat plate and then change to sphere

This seems to work for me. In my methods called by the sample_shape_ads_observer, I have noted that I don't need to do anything in the on_rename_workspace() method as `        # on_replace_workspace is also triggered by a workspace rename`. This was a slight oversight as the class didn't update it's`self.workspace_name` to the new name. So it was possible to get the plot out of sync by opening a sample shape plot, renaming the workspace and then running setsample again on the newly named workspace, but that works now. Tell me if this re-running of setsample works in all cases for you as it seems to work for me. I'd also note that the `        figure.canvas.draw()` calls are vital to the plot updating what is viewed. I think these calls are in the correct place.

DanielMurphy22 committed:
Requested Changes

Update docs with aspect, ortho and vertical axis commands
Actually rename workspace in sample_shape class on rename
Scale real lattice vectors to 80% length to make overlap clearer

DannyHindson reviewed (approved):
I can't reproduce the problem where the shape didn't update on running SetSample twice - not sure what I was up to before.
This all looks great now and will be a really useful new feature. Good job. Thanks for the changes!
